src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -1202,21 +1203,69 @@ private static int ApplyMask(byte[] toMask, int toMaskOffset, byte[] mask, int m
         /// <returns>The next index into the mask to be used for future applications of the mask.</returns>
         private static unsafe int ApplyMask(byte[] toMask, int toMaskOffset, int mask, int maskIndex, long count)
         {
-            Debug.Assert(toMaskOffset <= toMask.Length - count, $"Unexpected inputs: {toMaskOffset}, {toMask.Length}, {count}");
-            Debug.Assert(maskIndex < sizeof(int), $"Unexpected {nameof(maskIndex)}: {maskIndex}");
+            int maskShift = maskIndex * 8;
+            int shiftedMask = (int)(((uint)mask >> maskShift) | ((uint)mask << (32 - maskShift)));
+
+            // Try to use SIMD.  We can if the number of bytes we're trying to mask is at least as much
+            // as the width of a vector and if the width is an even multiple of the mask.
+            if (Vector.IsHardwareAccelerated &&
+                Vector<byte>.Count % sizeof(int) == 0 &&
+                count >= Vector<byte>.Count)
+            {
+                // Mask bytes a vector at a time.
+                Vector<byte> maskVector = Vector.AsVectorByte(new Vector<int>(shiftedMask));
+                while (count >= Vector<byte>.Count)
+                {
+                    (maskVector ^ new Vector<byte>(toMask, toMaskOffset)).CopyTo(toMask, toMaskOffset);
+                    toMaskOffset += Vector<byte>.Count;
+                    count -= Vector<byte>.Count;
+                }
+
+                // Fall through to processing any remaining bytes that were less than a vector width.
+                // Since we processed full masks at a time, we don't need to update maskIndex, and
+                // toMaskOffset has already been updated to point to the correct location.
+            }
 
-            byte* maskPtr = (byte*)&mask;
-            fixed (byte* toMaskPtr = toMask)
+            // If there are any bytes remaining (either we couldn't use vectors, or the count wasn't
+            // an even multiple of the vector width), process them without vectors.
+            if (count > 0)
             {
-                byte* p = toMaskPtr + toMaskOffset;
-                byte* end = p + count;
-                while (p < end)
+                fixed (byte* toMaskPtr = toMask)
                 {
-                    *p++ ^= maskPtr[maskIndex];
-                    maskIndex = (maskIndex + 1) & 3; // & 3 == faster % MaskLength
+                    // Get the location in the target array to continue processing.
+                    byte* p = toMaskPtr + toMaskOffset;
+
+                    // Try to go an int at a time if the remaining data is 4-byte aligned and there's enough remaining.
+                    if (((long)p % sizeof(int)) == 0)
+                    {
+                        while (count >= sizeof(int))
+                        {
+                            *((int*)p) ^= shiftedMask;
+                            p += sizeof(int);
+                            count -= sizeof(int);
```

Comment:
Can move this up two lines, immediately after while so its ready earlier in pipeline for next loop test